pkg/operator/controller/status.go

```diff
@@ -40,71 +46,153 @@ func (r *reconciler) syncOperatorStatus() error {
-		err = r.client.Status().Update(context.TODO(), co)
-		if err != nil {
-			return fmt.Errorf("failed to update clusteroperator %s: %v", co.Name, err)
-		}
+	ingressList := &operatorv1.IngressControllerList{}
+	if err := r.client.List(context.TODO(), ingressList, client.InNamespace(r.Namespace)); err != nil {
+		return nil, fmt.Errorf("failed to list IngressControllers: %v", err)
 	}
 
-	return nil
+	return ingressList.Items, nil
 }
 
-// getOperatorState gets and returns the resources necessary to compute the
-// operator's current state.
-func (r *reconciler) getOperatorState() (*corev1.Namespace, []operatorv1.IngressController, error) {
-	ns := manifests.RouterNamespace()
+// computeOperatorStatusVersions computes the operator's current versions and also updates progressing condition if needed.
+func (r *reconciler) computeOperatorStatusVersions(oldVersions []configv1.OperandVersion, conditions []configv1.ClusterOperatorStatusCondition) ([]configv1.OperandVersion, []configv1.ClusterOperatorStatusCondition) {
+	oldReleaseVersion := UnknownVersionValue
+	oldIngressControllerImage := UnknownVersionValue
+	for _, opv := range oldVersions {
+		switch opv.Name {
+		case OperatorVersionName:
+			oldReleaseVersion = opv.Version
+		case IngressControllerVersionName:
+			oldIngressControllerImage = opv.Version
+		}
+	}
 
-	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ns.Name}, ns); err != nil {
-		if errors.IsNotFound(err) {
-			return nil, nil, nil
+	operatorAvailable := false
+	operatorProgressing := false
+	var progressingCondition *configv1.ClusterOperatorStatusCondition
+	for i, cond := range conditions {
+		switch cond.Type {
+		case configv1.OperatorAvailable:
+			if cond.Status == configv1.ConditionTrue {
+				operatorAvailable = true
+			}
+		case configv1.OperatorProgressing:
+			progressingCondition = &conditions[i]
+			if cond.Status == configv1.ConditionTrue {
+				operatorProgressing = true
+			}
 		}
+	}
 
-		return nil, nil, fmt.Errorf(
-			"error getting Namespace %s: %v", ns.Name, err)
+	releaseVersion := r.OperatorReleaseVersion
+	ingressControllerImage := r.IngressControllerImage
+	// We need to report old version until the operator fully transitions to the new version.
+	// https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#version-reporting-during-an-upgrade
+	if !operatorAvailable || operatorProgressing {
+		releaseVersion = oldReleaseVersion
+		ingressControllerImage = oldIngressControllerImage
 	}
 
-	ingressList := &operatorv1.IngressControllerList{}
-	if err := r.client.List(context.TODO(), ingressList, client.InNamespace(r.Namespace)); err != nil {
-		return nil, nil, fmt.Errorf("failed to list IngressControllers: %v", err)
+	// Update progressing condition if needed
+	if (releaseVersion != r.OperatorReleaseVersion) || (ingressControllerImage != r.IngressControllerImage) {
+		if !operatorProgressing {
+			progressingCondition.Status = configv1.ConditionTrue
```

Comment:
Can you be sure `progressingCondition` is not nil?

Reply:
Yes, We populate progressing condition during the cluster operator creation so it should not be nil.

Reply:
What if something else created the clusteroperator, or modified it by deleting the condition? Seems like this would fail and crashloop.  Granted, it is an unlikely scenario.

Reply:
I see, we can assume that `computeOperatorStatusConditions` will have set the condition.  Never mind!